_posts/2020-02-17-eda-architecture.md

```diff
@@ -0,0 +1,181 @@
+
+So one year later, as our activity grew and development accelerated, we quickly realized that we needed:
+
+* maximum service decoupling;
+* easy distribution;
+* no brainer load balancing;
+* one to one, one to many, many to one, many to many asynchronous communication;
+* high availability: hot restart of services, transparent addition or removal of service instances (workers), resilience to (reasonable) downtime of some services;
+* support for heavy payloads (megabytes of mp3 audio);
+* no data loss, whatever happens.
+
+## Enter the event driven architecture
+
+The best way to achieve those goals is to free your mind from the classical pipeline point of view and instead see the value chain as an ecosystem of business services, each focused on providing a specific value and reacting to events (inputs) and producing new events (outputs). This new metaphor has not only technical benefits, but also business and organizational ones. By reasoning in terms of business units of your value chain, its easier to identify the people involved, the business experts who are the references for the job, the exact value added by the service, etc…
+
+Events are precisely defined messages that streams on a message bus, and each logical service (implementing one such business service as explained above) subscribes to the events that are relevant to it, without needing any knowledge about what produced them and how. They also push their own events on the bus, without caring about what consumes them.
+
+In that way, we completely decouple the services between each other and the message broker running the message bus provides us with load balancing, distribution and high availability for free. For that purpose, we chose [RabbitMQ](https://www.rabbitmq.com/) as message broker for its reliability record and ease of use.
+
+Now, the messages *are* the API, the only business and technical reference.
+
+After much thinking and experiments, we came with these core design principles that are very important, and after 4 months of production use, we are very glad we complied with them from the start:
+
+### Business Services and Data Processing Services
+
+We make a distinction between Business Services and Data Processing Services (a.k.a. utility services) to cleanly separate business logic from data processing complexity.
+
+DP Services are expected to be  **pure, stateless, services** that provide some kind of algorithmic data processing (computations, transformations, …). Moreover, they are also context free: they should not depend on business knowledge, assumptions or external data sources. All they need to do their processing must be in the command itself. They should not have to query a tier to get more data. DP services are kind of *universal* libraries and can even be provided by tiers.
```

Comment:
By business knowledge, you mean business rules?

Reply:
Business rules or assumptions or any business specific interpretation of data.  Should I detail these in the article to make the point clearer?

Reply:
"business rules" is easier to understand. I don't think we should detail more